lib/ex_machina/ecto_strategy.ex

```diff
@@ -23,4 +25,71 @@ defmodule ExMachina.EctoStrategy do
   def handle_insert(_record, _opts) do
     raise "expected :repo to be given to ExMachina.EctoStrategy"
   end
+
+  defp cast(record) do
+    record
+    |> cast_fields
+    |> cast_assocs
+  end
+
+  defp cast_fields(struct) do
+    struct
+    |> ExMachina.Ecto.drop_ecto_fields
+    |> Map.keys
+    |> cast_fields(struct)
+  end
+
+  defp cast_fields(fields, struct) do
+    Enum.reduce(fields, struct, fn(field, struct) ->
+      casted_value = cast_field(field, struct)
+      Map.put(struct, field, casted_value)
+    end)
+  end
+
+  defp cast_field(field, %{__struct__: schema} = struct) do
+    field_type = schema.__schema__(:type, field)
+    value = Map.get(struct, field)
+
+    if field_type do
+      cast_value(field_type, value, struct)
+    else
+      value
+    end
+  end
+
+  defp cast_value(field_type, value, struct) do
+    case Ecto.Type.cast(field_type, value) do
+      {:ok, value} ->
+        value
+      _ ->
+        raise "Failed to cast `#{value}` of type #{field_type} in #{inspect struct}."
+    end
+  end
+
+  defp cast_assocs(%{__struct__: schema} = struct) do
+    assocs = schema.__schema__(:associations)
+
+    Enum.reduce(assocs, struct, fn(assoc, struct) ->
+      casted_value = cast_assoc(assoc, struct)
+      Map.put(struct, assoc, casted_value)
+    end)
+  end
+
+  defp cast_assoc(assoc, %{__struct__: schema} = struct) do
+    original_assoc = Map.get(struct, assoc)
+
+    case original_assoc do
+      %{__meta__: %{__struct__: Ecto.Schema.Metadata, state: :built}} ->
+        cast(original_assoc)
+      %{__struct__: _} ->
+        original_assoc
+      records = [_ | _] ->
+        Enum.map(records, &(cast(&1)))
+      %{} ->
+        assoc_type = schema.__schema__(:association, assoc).related
+        assoc_type |> struct |> Map.merge(original_assoc) |> cast
+      _ ->
```

Comment:
Out of curiosity, what would the underscore be. Something like a bare int or a string or something?